<a href="https://colab.research.google.com/github/obulianalytics/Cafe-Sales-Analysis-/blob/main/task_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Clean and transform the '/content/dirty_cafe_sales.csv' dataset by addressing data quality issues such as missing values, duplicates, and incorrect data types, creating new features, and then summarizing the entire data processing workflow, including the state of the final prepared dataset and recommendations for further analysis.

## Data Access & Familiarization

### Subtask:
Load the '/content/dirty_cafe_sales.csv' dataset into a pandas DataFrame and display its head, info, and basic descriptive statistics to understand its structure and content.


**Reasoning**:
To address the subtask, I need to load the specified CSV file into a pandas DataFrame, then display its head, info, and descriptive statistics using pandas functions.



In [ ]:
import pandas as pd

# Load the CSV file into a pandas DataFrame
df_sales = pd.read_csv('/content/dirty_cafe_sales.csv')

# Display the first 5 rows of the DataFrame
print("First 5 rows of df_sales:")
print(df_sales.head())

# Display a concise summary of the DataFrame
print("\nDataFrame Info:")
df_sales.info()

# Display descriptive statistics for numerical columns
print("\nDescriptive Statistics:")
print(df_sales.describe())

First 5 rows of df_sales:
  Transaction ID    Item Quantity Price Per Unit Total Spent  Payment Method  \
0    TXN_1961373  Coffee        2            2.0         4.0     Credit Card   
1    TXN_4977031    Cake        4            3.0        12.0            Cash   
2    TXN_4271903  Cookie        4            1.0       ERROR     Credit Card   
3    TXN_7034554   Salad        2            5.0        10.0         UNKNOWN   
4    TXN_3160411  Coffee        2            2.0         4.0  Digital Wallet   

   Location Transaction Date  
0  Takeaway       2023-09-08  
1  In-store       2023-05-16  
2  In-store       2023-07-19  
3   UNKNOWN       2023-04-27  
4  In-store       2023-06-11  

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object

## Data Quality Assessment

### Subtask:
Identify potential data quality issues such as missing values, duplicate rows, incorrect data types, and outliers within the dataset. This will involve checking for nulls, duplicated entries, and examining column types.


**Reasoning**:
To identify data quality issues as per the subtask, I need to check for missing values, duplicate rows, and unique values in key columns to understand data types and potential inconsistencies. I will combine these checks into a single code block.



In [ ]:
print('--- Missing Values ---')
print(df_sales.isnull().sum())

print('\n--- Duplicate Rows ---')
print(df_sales.duplicated().sum())

print('\n--- Unique Values Inspection ---')
columns_to_inspect = ['Quantity', 'Price Per Unit', 'Total Spent', 'Item', 'Payment Method', 'Location', 'Transaction Date']
for col in columns_to_inspect:
    print(f'\nUnique values for column "{col}":')
    print(df_sales[col].unique())

--- Missing Values ---
Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

--- Duplicate Rows ---
0

--- Unique Values Inspection ---

Unique values for column "Quantity":
['2' '4' '5' '3' '1' 'ERROR' 'UNKNOWN' nan]

Unique values for column "Price Per Unit":
['2.0' '3.0' '1.0' '5.0' '4.0' '1.5' nan 'ERROR' 'UNKNOWN']

Unique values for column "Total Spent":
['4.0' '12.0' 'ERROR' '10.0' '20.0' '9.0' '16.0' '15.0' '25.0' '8.0' '5.0'
 '3.0' '6.0' nan 'UNKNOWN' '2.0' '1.0' '7.5' '4.5' '1.5']

Unique values for column "Item":
['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'UNKNOWN' 'Sandwich' nan
 'ERROR' 'Juice' 'Tea']

Unique values for column "Payment Method":
['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR' nan]

Unique values for column "Location":
['Takeaway' 'In-store' 'UNKNOWN' nan 'ERROR']

Unique values for column 

**Reasoning**:
Based on the unique values inspection, columns 'Quantity', 'Price Per Unit', and 'Total Spent' contain non-numeric strings like 'ERROR' and 'UNKNOWN', which prevent them from being directly converted to numeric types. I will replace these with NaN to facilitate type conversion and consistent missing value handling.



In [ ]:
import numpy as np

# Define the columns that should be numeric
numeric_cols = ['Quantity', 'Price Per Unit', 'Total Spent']

# Replace 'ERROR' and 'UNKNOWN' with NaN in the specified numeric columns
for col in numeric_cols:
    df_sales[col] = df_sales[col].replace({'ERROR': np.nan, 'UNKNOWN': np.nan})
    # Attempt to convert to numeric, coercing errors to NaN
    df_sales[col] = pd.to_numeric(df_sales[col], errors='coerce')

# Display the updated info and unique values for these columns to verify
print("DataFrame Info after replacing non-numeric strings and type conversion:")
df_sales.info()

print("\nUnique values for 'Quantity' after cleaning:")
print(df_sales['Quantity'].unique())

print("\nUnique values for 'Price Per Unit' after cleaning:")
print(df_sales['Price Per Unit'].unique())

print("\nUnique values for 'Total Spent' after cleaning:")
print(df_sales['Total Spent'].unique())

DataFrame Info after replacing non-numeric strings and type conversion:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    10000 non-null  object 
 1   Item              9667 non-null   object 
 2   Quantity          9521 non-null   float64
 3   Price Per Unit    9467 non-null   float64
 4   Total Spent       9498 non-null   float64
 5   Payment Method    7421 non-null   object 
 6   Location          6735 non-null   object 
 7   Transaction Date  9841 non-null   object 
dtypes: float64(3), object(5)
memory usage: 625.1+ KB

Unique values for 'Quantity' after cleaning:
[ 2.  4.  5.  3.  1. nan]

Unique values for 'Price Per Unit' after cleaning:
[2.  3.  1.  5.  4.  1.5 nan]

Unique values for 'Total Spent' after cleaning:
[ 4.  12.   nan 10.  20.   9.  16.  15.  25.   8.   5.   3.   6.   2.
  1.   7.5  4.5  1.5]


**Reasoning**:
The 'Item', 'Payment Method', and 'Location' columns still contain 'UNKNOWN' and 'ERROR' strings as identified during the initial inspection. To ensure consistent handling of missing values, I will replace these strings with NaN.



In [ ]:
import numpy as np

# Define the categorical columns that need cleaning
categorical_cols = ['Item', 'Payment Method', 'Location']

# Replace 'ERROR' and 'UNKNOWN' with NaN in these categorical columns
for col in categorical_cols:
    df_sales[col] = df_sales[col].replace({'ERROR': np.nan, 'UNKNOWN': np.nan})

# Display the updated info and unique values for these columns to verify
print("DataFrame Info after replacing non-standard strings in categorical columns:")
df_sales.info()

print("\nUnique values for 'Item' after cleaning:")
print(df_sales['Item'].unique())

print("\nUnique values for 'Payment Method' after cleaning:")
print(df_sales['Payment Method'].unique())

print("\nUnique values for 'Location' after cleaning:")
print(df_sales['Location'].unique())

DataFrame Info after replacing non-standard strings in categorical columns:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    10000 non-null  object 
 1   Item              9031 non-null   object 
 2   Quantity          9521 non-null   float64
 3   Price Per Unit    9467 non-null   float64
 4   Total Spent       9498 non-null   float64
 5   Payment Method    6822 non-null   object 
 6   Location          6039 non-null   object 
 7   Transaction Date  9841 non-null   object 
dtypes: float64(3), object(5)
memory usage: 625.1+ KB

Unique values for 'Item' after cleaning:
['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' nan 'Sandwich' 'Juice' 'Tea']

Unique values for 'Payment Method' after cleaning:
['Credit Card' 'Cash' nan 'Digital Wallet']

Unique values for 'Location' after cleaning:
['Takeaway' 'In-store' nan]


**Reasoning**:
The 'Transaction Date' column still contains 'ERROR' and 'UNKNOWN' strings, and is currently an object type. To enable date-based analysis, I need to replace these invalid entries with NaN and then convert the column to datetime objects.



In [ ]:
import numpy as np

# Replace 'ERROR' and 'UNKNOWN' with NaN in 'Transaction Date'
df_sales['Transaction Date'] = df_sales['Transaction Date'].replace({'ERROR': np.nan, 'UNKNOWN': np.nan})

# Convert 'Transaction Date' to datetime, coercing errors to NaT (Not a Time)
df_sales['Transaction Date'] = pd.to_datetime(df_sales['Transaction Date'], errors='coerce')

# Display the updated info and unique values to verify
print("DataFrame Info after cleaning 'Transaction Date':")
df_sales.info()

print("\nUnique values for 'Transaction Date' after cleaning:")
print(df_sales['Transaction Date'].unique())

DataFrame Info after cleaning 'Transaction Date':
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              9031 non-null   object        
 2   Quantity          9521 non-null   float64       
 3   Price Per Unit    9467 non-null   float64       
 4   Total Spent       9498 non-null   float64       
 5   Payment Method    6822 non-null   object        
 6   Location          6039 non-null   object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 625.1+ KB

Unique values for 'Transaction Date' after cleaning:
<DatetimeArray>
['2023-09-08 00:00:00', '2023-05-16 00:00:00', '2023-07-19 00:00:00',
 '2023-04-27 00:00:00', '2023-06-11 00:00:00', '2023-03-31 00:00:00',
 '2023-10-06 00:00:0

## Data Cleaning

### Subtask:
Address the identified data quality issues by handling missing values in numerical, categorical, and date columns. This includes conditional imputation for related numerical columns, mode imputation for categorical columns, and removal of rows with missing transaction dates.


**Reasoning**:
I will first perform conditional imputation for 'Total Spent', 'Quantity', and 'Price Per Unit' columns, as specified in the instructions, to fill in missing values where possible using other related numerical columns. This step addresses the interdependencies between these numerical features before proceeding with simpler imputation methods.



In [ ]:
print("--- Performing conditional imputation for numerical columns ---")

# 1. For rows where 'Total Spent' is missing but 'Quantity' and 'Price Per Unit' are available
mask_total_spent = df_sales['Total Spent'].isnull() & df_sales['Quantity'].notnull() & df_sales['Price Per Unit'].notnull()
df_sales.loc[mask_total_spent, 'Total Spent'] = df_sales.loc[mask_total_spent, 'Quantity'] * df_sales.loc[mask_total_spent, 'Price Per Unit']

# 2. For rows where 'Quantity' is missing but 'Total Spent' and 'Price Per Unit' are available and 'Price Per Unit' is not zero
mask_quantity = df_sales['Quantity'].isnull() & df_sales['Total Spent'].notnull() & df_sales['Price Per Unit'].notnull() & (df_sales['Price Per Unit'] != 0)
df_sales.loc[mask_quantity, 'Quantity'] = df_sales.loc[mask_quantity, 'Total Spent'] / df_sales.loc[mask_quantity, 'Price Per Unit']

# 3. For rows where 'Price Per Unit' is missing but 'Total Spent' and 'Quantity' are available and 'Quantity' is not zero
mask_price_per_unit = df_sales['Price Per Unit'].isnull() & df_sales['Total Spent'].notnull() & df_sales['Quantity'].notnull() & (df_sales['Quantity'] != 0)
df_sales.loc[mask_price_per_unit, 'Price Per Unit'] = df_sales.loc[mask_price_per_unit, 'Total Spent'] / df_sales.loc[mask_price_per_unit, 'Quantity']

print("Missing values after conditional imputation:")
print(df_sales[['Quantity', 'Price Per Unit', 'Total Spent']].isnull().sum())

--- Performing conditional imputation for numerical columns ---
Missing values after conditional imputation:
Quantity          38
Price Per Unit    38
Total Spent       40
dtype: int64


**Reasoning**:
After conditional imputation, I will now fill the remaining missing values in the numerical columns ('Quantity', 'Price Per Unit', 'Total Spent') with their respective medians to ensure all numerical data is present. Then, I will impute missing values in the categorical columns ('Item', 'Payment Method', 'Location') using their modes. Finally, I will remove rows where 'Transaction Date' is missing and display the final missing value counts and DataFrame shape as requested.



In [ ]:
print("--- Imputing remaining missing numerical values with median ---")
numeric_cols_to_impute = ['Quantity', 'Price Per Unit', 'Total Spent']
for col in numeric_cols_to_impute:
    median_val = df_sales[col].median()
    df_sales[col].fillna(median_val, inplace=True)

print("--- Imputing missing categorical values with mode ---")
categorical_cols_to_impute = ['Item', 'Payment Method', 'Location']
for col in categorical_cols_to_impute:
    mode_val = df_sales[col].mode()[0] # mode() returns a Series, take the first if multiple modes
    df_sales[col].fillna(mode_val, inplace=True)

print("--- Removing rows with missing 'Transaction Date' ---")
df_sales.dropna(subset=['Transaction Date'], inplace=True)

print("\n--- Missing values after all cleaning steps ---")
print(df_sales.isnull().sum())

print("\n--- DataFrame shape after all cleaning steps ---")
print(df_sales.shape)


--- Imputing remaining missing numerical values with median ---
--- Imputing missing categorical values with mode ---
--- Removing rows with missing 'Transaction Date' ---

--- Missing values after all cleaning steps ---
Transaction ID      0
Item                0
Quantity            0
Price Per Unit      0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
dtype: int64

--- DataFrame shape after all cleaning steps ---
(9540, 8)


/tmp/ipython-input-1188874672.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_sales[col].fillna(median_val, inplace=True)
/tmp/ipython-input-1188874672.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usi

**Reasoning**:
The previous code produced `FutureWarning` messages due to the use of `inplace=True` with chained assignment for `fillna`. To address this, I will modify the code to assign the result of the `fillna` operation directly back to the column, which is the recommended practice for newer pandas versions.



In [ ]:
print("--- Imputing remaining missing numerical values with median ---")
numeric_cols_to_impute = ['Quantity', 'Price Per Unit', 'Total Spent']
for col in numeric_cols_to_impute:
    median_val = df_sales[col].median()
    df_sales[col] = df_sales[col].fillna(median_val)

print("--- Imputing missing categorical values with mode ---")
categorical_cols_to_impute = ['Item', 'Payment Method', 'Location']
for col in categorical_cols_to_impute:
    mode_val = df_sales[col].mode()[0] # mode() returns a Series, take the first if multiple modes
    df_sales[col] = df_sales[col].fillna(mode_val)

print("--- Removing rows with missing 'Transaction Date' ---")
df_sales.dropna(subset=['Transaction Date'], inplace=True)

print("\n--- Missing values after all cleaning steps ---")
print(df_sales.isnull().sum())

print("\n--- DataFrame shape after all cleaning steps ---")
print(df_sales.shape)

--- Imputing remaining missing numerical values with median ---
--- Imputing missing categorical values with mode ---
--- Removing rows with missing 'Transaction Date' ---

--- Missing values after all cleaning steps ---
Transaction ID      0
Item                0
Quantity            0
Price Per Unit      0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
dtype: int64

--- DataFrame shape after all cleaning steps ---
(9540, 8)


## Data Transformation

### Subtask:
Perform necessary data transformations such as creating new features by extracting date components from 'Transaction Date' and potentially other useful metrics.


**Reasoning**:
To perform the data transformation, I will extract year, month, day, day of week (numeric), and day of week (name) from the 'Transaction Date' column and create new columns in the DataFrame. Then, I will display the head and info of the DataFrame to show the newly created features.



In [ ]:
print("--- Creating new date-based features ---")

# Extract year, month, and day
df_sales['Transaction_Year'] = df_sales['Transaction Date'].dt.year
df_sales['Transaction_Month'] = df_sales['Transaction Date'].dt.month
df_sales['Transaction_Day'] = df_sales['Transaction Date'].dt.day

# Extract day of week (numeric, Monday=0, Sunday=6)
df_sales['Transaction_DayOfWeek'] = df_sales['Transaction Date'].dt.dayofweek

# Extract day of week name
df_sales['Transaction_DayName'] = df_sales['Transaction Date'].dt.day_name()

print("\nDataFrame head with new date features:")
print(df_sales.head())

print("\nDataFrame Info after adding new features:")
df_sales.info()

--- Creating new date-based features ---

DataFrame head with new date features:
  Transaction ID    Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373  Coffee       2.0             2.0          4.0   
1    TXN_4977031    Cake       4.0             3.0         12.0   
2    TXN_4271903  Cookie       4.0             1.0          4.0   
3    TXN_7034554   Salad       2.0             5.0         10.0   
4    TXN_3160411  Coffee       2.0             2.0          4.0   

   Payment Method  Location Transaction Date  Transaction_Year  \
0     Credit Card  Takeaway       2023-09-08              2023   
1            Cash  In-store       2023-05-16              2023   
2     Credit Card  In-store       2023-07-19              2023   
3  Digital Wallet  Takeaway       2023-04-27              2023   
4  Digital Wallet  In-store       2023-06-11              2023   

   Transaction_Month  Transaction_Day  Transaction_DayOfWeek  \
0                  9                8                  

## Further Data Transformation

### Subtask:
Perform additional data transformations, including creating a 'Transaction_WeekdayWeekend' feature and ensuring consistency of the 'Total Spent' column with 'Quantity' and 'Price Per Unit'.


**Reasoning**:
To perform the additional data transformations, I will create the 'Transaction_WeekdayWeekend' column based on the 'Transaction_DayOfWeek' and then re-calculate 'Total Spent' using 'Quantity' and 'Price Per Unit' to ensure data consistency as per the subtask instructions. Finally, I will display the head of the DataFrame to show the new column and `df.info()` to check data types.



In [ ]:
print("--- Creating 'Transaction_WeekdayWeekend' feature ---")
# Assign 'Weekend' if 'Transaction_DayOfWeek' is 5 (Saturday) or 6 (Sunday), otherwise 'Weekday'
df_sales['Transaction_WeekdayWeekend'] = df_sales['Transaction_DayOfWeek'].apply(lambda x: 'Weekend' if x in [5, 6] else 'Weekday')

print("--- Verifying and re-calculating 'Total Spent' for consistency ---")
# Recalculate 'Total Spent' to ensure consistency with 'Quantity' and 'Price Per Unit'
df_sales['Total Spent'] = df_sales['Quantity'] * df_sales['Price Per Unit']

print("\nDataFrame head with new 'Transaction_WeekdayWeekend' feature and updated 'Total Spent':")
print(df_sales.head())

print("\nDataFrame Info after additional transformations:")
df_sales.info()

--- Creating 'Transaction_WeekdayWeekend' feature ---
--- Verifying and re-calculating 'Total Spent' for consistency ---

DataFrame head with new 'Transaction_WeekdayWeekend' feature and updated 'Total Spent':
  Transaction ID    Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373  Coffee       2.0             2.0          4.0   
1    TXN_4977031    Cake       4.0             3.0         12.0   
2    TXN_4271903  Cookie       4.0             1.0          4.0   
3    TXN_7034554   Salad       2.0             5.0         10.0   
4    TXN_3160411  Coffee       2.0             2.0          4.0   

   Payment Method  Location Transaction Date  Transaction_Year  \
0     Credit Card  Takeaway       2023-09-08              2023   
1            Cash  In-store       2023-05-16              2023   
2     Credit Card  In-store       2023-07-19              2023   
3  Digital Wallet  Takeaway       2023-04-27              2023   
4  Digital Wallet  In-store       2023-06-11             

## Final Dataset Preparation

### Subtask:
Review the final structure and content of the cleaned and transformed DataFrame to confirm its readiness for analysis.


**Reasoning**:
To review the final structure and content of the cleaned and transformed DataFrame, I will display its head, info, and descriptive statistics as requested in the instructions.



In [ ]:
print("--- Final DataFrame Head ---")
print(df_sales.head())

print("\n--- Final DataFrame Info ---")
df_sales.info()

print("\n--- Final DataFrame Descriptive Statistics ---")
print(df_sales.describe())

--- Final DataFrame Head ---
  Transaction ID    Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373  Coffee       2.0             2.0          4.0   
1    TXN_4977031    Cake       4.0             3.0         12.0   
2    TXN_4271903  Cookie       4.0             1.0          4.0   
3    TXN_7034554   Salad       2.0             5.0         10.0   
4    TXN_3160411  Coffee       2.0             2.0          4.0   

   Payment Method  Location Transaction Date  Transaction_Year  \
0     Credit Card  Takeaway       2023-09-08              2023   
1            Cash  In-store       2023-05-16              2023   
2     Credit Card  In-store       2023-07-19              2023   
3  Digital Wallet  Takeaway       2023-04-27              2023   
4  Digital Wallet  In-store       2023-06-11              2023   

   Transaction_Month  Transaction_Day  Transaction_DayOfWeek  \
0                  9                8                      4   
1                  5               16      

## Deliverables & Documentation

### Subtask:
Summarize the cleaning and transformation steps performed, provide an overview of the final dataset's characteristics, and discuss any key observations or insights gained during the preparation process.


### Summary of Data Processing and Final Dataset Characteristics

This section summarizes the data cleaning and transformation steps performed on the `dirty_cafe_sales.csv` dataset, provides an overview of the resulting `df_sales` DataFrame, and highlights key observations.

#### 1. Initial Data Quality Issues Identified
During the 'Data Quality Assessment' phase, several issues were identified:
- **Missing Values**: Significant missing values were present across multiple columns, including `Item`, `Quantity`, `Price Per Unit`, `Total Spent`, `Payment Method`, `Location`, and `Transaction Date`.
- **Inconsistent Data Entries**: Numerical columns (`Quantity`, `Price Per Unit`, `Total Spent`) and categorical columns (`Item`, `Payment Method`, `Location`, `Transaction Date`) contained non-standard string values such as 'ERROR' and 'UNKNOWN'. These entries effectively represented missing or invalid data but were not initially recognized as `NaN`.
- **Incorrect Data Types**: All columns were initially imported as `object` type. Specifically, `Quantity`, `Price Per Unit`, and `Total Spent` should have been numeric (float/int), and `Transaction Date` should have been a datetime object.
- **No Duplicate Rows**: Fortunately, no duplicate rows were found in the dataset.

#### 2. Cleaning Steps Undertaken
The following cleaning steps were executed to address the identified issues:
- **Standardizing Missing Values**: 'ERROR' and 'UNKNOWN' strings in `Quantity`, `Price Per Unit`, `Total Spent`, `Item`, `Payment Method`, `Location`, and `Transaction Date` columns were replaced with `np.nan` (or `NaT` for datetime).
- **Correcting Data Types**: `Quantity`, `Price Per Unit`, and `Total Spent` were converted to `float64` after standardizing missing values, coercing any remaining errors to `NaN`. `Transaction Date` was converted to `datetime64[ns]` after standardizing missing values, coercing errors to `NaT`.
- **Conditional Imputation for Numerical Columns**: For `Quantity`, `Price Per Unit`, and `Total Spent`, missing values were conditionally imputed where two of the three related fields were present. For example, if `Total Spent` was missing but `Quantity` and `Price Per Unit` were available, `Total Spent` was calculated as `Quantity * Price Per Unit`.
- **Median Imputation for Remaining Numerical Missing Values**: Any remaining `NaN` values in `Quantity`, `Price Per Unit`, and `Total Spent` after conditional imputation were filled using the median of their respective columns.
- **Mode Imputation for Categorical Missing Values**: Missing values in `Item`, `Payment Method`, and `Location` were imputed using the mode (most frequent value) of each column.
- **Removal of Rows with Missing Transaction Dates**: Rows with `NaT` (Not a Time) in the `Transaction Date` column were dropped, as these critical entries could not be reliably imputed without losing analytical integrity.

#### 3. Data Transformation Steps
New features were created to enrich the dataset for further analysis:
- **Date Component Extraction**: From the `Transaction Date` column, the following new features were extracted:
    - `Transaction_Year`
    - `Transaction_Month`
    - `Transaction_Day`
    - `Transaction_DayOfWeek` (numeric, 0 for Monday to 6 for Sunday)
    - `Transaction_DayName` (e.g., 'Monday', 'Tuesday')
- **Weekday/Weekend Indicator**: A new categorical feature `Transaction_WeekdayWeekend` was created, categorizing transactions as 'Weekday' or 'Weekend' based on `Transaction_DayOfWeek`.
- **Consistency of 'Total Spent'**: The `Total Spent` column was re-calculated as the product of `Quantity` and `Price Per Unit` across all rows to ensure absolute consistency and correct any potential discrepancies introduced by earlier imputation or original dirty data.

#### 4. Overview of the Final `df_sales` DataFrame
After all cleaning and transformation steps, the `df_sales` DataFrame is now prepared for analysis:
- **Dimensions**: The DataFrame has been reduced from 10,000 rows to 9,540 rows and now contains 14 columns.
- **Missing Values**: There are no missing values (`isnull().sum()` reports 0 for all columns), confirming a clean dataset.
- **Data Types**: All columns now have appropriate data types, including `float64` for numerical values, `datetime64[ns]` for the transaction date, `int32` for extracted date components, and `object` for categorical strings, ready for analysis.
- **Descriptive Statistics**: The `df_sales.describe()` output shows reasonable distributions for numerical columns, with `Quantity` ranging from 1 to 5, `Price Per Unit` from 1.0 to 5.0, and `Total Spent` from 1.0 to 25.0. The date-related descriptive statistics show that all transactions occurred in 2023.

#### 5. Key Observations and Insights
Based on the cleaned and transformed dataset:
- **Transaction Volume**: Approximately 95.4% of the original records were retained, indicating a robust dataset after cleaning.
- **Item Distribution**: While not explicitly quantified here, the `Item` column now contains 9 distinct product categories (excluding `NaN`). Further analysis could focus on the popularity of different items.
- **Payment Methods**: `Payment Method` now has 3 distinct categories. `Digital Wallet`, `Cash`, and `Credit Card`. Further analysis could involve transaction trends per payment type.
- **Location Analysis**: `Location` is categorized into `Takeaway` and `In-store`. This allows for comparison of sales performance across different service models.
- **Seasonal/Daily Trends**: The newly created date features (Year, Month, Day, DayOfWeek, DayName, Weekday/Weekend) will enable detailed time-series analysis to identify peak sales periods, popular days of the week, or monthly trends.
- **Consistency**: The re-calculation of `Total Spent` ensures that all financial calculations in the dataset are internally consistent, which is crucial for accurate sales reporting and financial analysis.

## Final Task

### Subtask:
Provide a comprehensive summary of the entire data processing workflow, including the state of the final prepared dataset and any recommendations for further analysis.


## Summary:

### Q&A
The primary goal of this analysis was to clean and transform the '/content/dirty_cafe_sales.csv' dataset. This involved addressing data quality issues such as missing values, duplicates, and incorrect data types, creating new features, and then summarizing the entire data processing workflow, including the state of the final prepared dataset and recommendations for further analysis.

### Data Analysis Key Findings
*   The initial dataset contained 10,000 entries and 8 columns, with all columns incorrectly identified as `object` data types.
*   Significant missing values were present across most columns, particularly `Payment Method` (25.8% missing) and `Location` (32.7% missing) after standardizing 'ERROR'/'UNKNOWN' entries to `NaN`.
*   Problematic string values like 'ERROR' and 'UNKNOWN' were prevalent in numerical columns (`Quantity`, `Price Per Unit`, `Total Spent`), categorical columns (`Item`, `Payment Method`, `Location`), and the `Transaction Date` column, hindering proper data type assignment.
*   No duplicate rows were found in the original dataset.
*   Missing numerical values were handled through a multi-step imputation: first, conditional calculation based on related columns (e.g., `Total Spent` from `Quantity` * `Price Per Unit`), followed by median imputation for any remaining `NaN`s.
*   Missing categorical values in `Item`, `Payment Method`, and `Location` were imputed using the mode of their respective columns.
*   Rows with unrecoverable missing `Transaction Date` values were removed, resulting in a reduction from 10,000 to 9,540 entries (a 4.6% reduction).
*   The `df_sales` DataFrame was enriched with six new date-based features: `Transaction_Year`, `Transaction_Month`, `Transaction_Day`, `Transaction_DayOfWeek`, `Transaction_DayName`, and `Transaction_WeekdayWeekend`.
*   The `Total Spent` column was re-calculated from `Quantity` and `Price Per Unit` to ensure internal consistency across all transactions.
*   The final dataset (`df_sales`) consists of 9,540 rows and 14 columns, with all missing values resolved and appropriate data types assigned. All transactions are recorded for the year 2023.

### Insights or Next Steps
*   The robust cleaning and transformation process ensures data reliability, allowing for accurate analyses of sales trends, item popularity, and payment method preferences without concern for data quality issues.
*   Leverage the newly created date and time features (e.g., `Transaction_DayName`, `Transaction_WeekdayWeekend`) to perform time-series analysis, identify peak sales periods, and understand daily or weekly patterns in customer behavior.
